File: src/python/day08.py
import math

example_input = """
162,817,812
57,618,57
906,360,560
592,479,940
352,342,300
466,668,158
542,29,236
431,825,988
739,650,466
52,470,668
216,146,977
819,987,18
117,168,530
805,96,715
346,949,466
970,615,88
941,993,340
862,61,35
984,92,344
425,690,689
"""

def parse_input(input):
    return [list(map(int, line.split(","))) for line in input.strip().split("\n")]

def count_connected_components(graph) -> int:
    visited = set()
    def dfs(node):
        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                dfs(neighbor)

    count = []
    for i in range(len(graph)):
        if i not in visited:
            visited_before_size = len(visited)
            visited.add(i)
            dfs(i)
            visited_after_size = len(visited)
            count.append(visited_after_size - visited_before_size)
    return math.prod(sorted(count)[-3:])

def calculate_joined_graph(graph, max_joints) -> tuple[dict[int, list[int]], tuple[int, int]]:
  distances = {}
  for i in range(len(graph)):
      for j in range(i + 1, len(graph)):
          distances[(i, j)] = math.sqrt((graph[i][0] - graph[j][0]) ** 2 + (graph[i][1] - graph[j][1]) ** 2 + (graph[i][2] - graph[j][2]) ** 2)
  distances = sorted(distances.items(), key=lambda x: x[1])

  joined_graph = [ [] for _ in range(len(graph))]
  joints = 0
  in_graph = set()
  last_joints = None
  for (i, j), d in distances:
      if joints >= max_joints:
          break
      if i not in in_graph or j not in in_graph:
          last_joints = (i, j)
      in_graph.add(i)
      in_graph.add(j)
      joined_graph[i].append(j)
      joined_graph[j].append(i)
      joints += 1
  return joined_graph, last_joints


def part1(input):
    joined_graph, last_joints = calculate_joined_graph(input, max_joints=1000)
    return count_connected_components(joined_graph)

def part2(input):
    joined_graph, last_joints = calculate_joined_graph(input, max_joints=1000000)
    return input[last_joints[0]][0] * input[last_joints[1]][0]


input = open("input/day08.txt", "r").read()
# input = example_input.strip()
print("PART 1 =", part1(parse_input(input)))
print("PART 2 =", part2(parse_input(input)))

File: src/rust/day08.rs
use itertools::Itertools;
use petgraph::graph::UnGraph;
use petgraph::{algo::tarjan_scc, graph::NodeIndex};
use std::collections::HashSet;

type Point = [i64; 3];

fn parse_input(input: &str) -> Vec<Point> {
    input
        .lines()
        .map(|line| {
            line.split(',')
                .map(|n| n.trim().parse().unwrap())
                .collect::<Vec<i64>>()
                .try_into()
                .unwrap()
        })
        .collect()
}

fn join_graph(points: &[Point], max_joints: Option<usize>) -> (UnGraph<usize, ()>, Option<(usize, usize)>) {
    let mut edges = (0..points.len())
        .tuple_combinations()
        .map(|(i, j)| {
            let (p1, p2) = (points[i], points[j]);
            let dist =
                ((p1[0] - p2[0]).pow(2) as f64 + (p1[1] - p2[1]).pow(2) as f64 + (p1[2] - p2[2]).pow(2) as f64).sqrt();
            (i, j, dist)
        })
        .collect::<Vec<_>>();

    edges.sort_by(|a, b| a.2.partial_cmp(&b.2).unwrap());

    let mut graph = UnGraph::<usize, ()>::new_undirected();
    for i in 0..points.len() {
        graph.add_node(i);
    }

    let mut in_graph = HashSet::new();
    let mut last_joints = None;
    for (i, j, _) in edges.into_iter().take(max_joints.unwrap_or(usize::MAX)) {
        if !in_graph.contains(&i) || !in_graph.contains(&j) {
            last_joints = Some((i, j));
        }
        in_graph.insert(i);
        in_graph.insert(j);
        graph.add_edge(NodeIndex::new(i), NodeIndex::new(j), ());
    }

    (graph, last_joints)
}

pub fn count_connected_components(input: &str) -> usize {
    let points = parse_input(input);
    let (joint_graph, _) = join_graph(&points, Some(1000));
    tarjan_scc(&joint_graph)
        .into_iter()
        .map(|comp| comp.len())
        .sorted()
        .rev()
        .take(3)
        .product()
}

pub fn count_last_joint(input: &str) -> i64 {
    let points = parse_input(input);
    if let (_, Some((u, v))) = join_graph(&points, None) {
        points[u][0] * points[v][0]
    } else {
        unreachable!()
    }
}

pub fn main() {
    let input = std::fs::read_to_string("input/day08.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", count_connected_components(&input));
    println!("PART 2 = {}", count_last_joint(&input));
    println!("Execution time: {:?}", now.elapsed());
}
